### PR TITLE
v05 misc fixes

### PR DESCRIPTION
**Pull request by ziofil**
**Context:**
Various improvements

**Description of the Change:**
- fixed double bar when optimizing
- removed placeholder at beginning of opt_history
- fixed tests accordingly
- we now have rich tracebacks :)

### CODE REVIEW

File: tests/test_training/test_opt.py [hunk 1]
```diff
@@ -107,16 +108,16 @@ def cost_fn():
     )
     assert np.allclose(np.cos(bs.theta.value) ** 2, k / (i + k), atol=1e-2)
     assert "null_cb" in opt.callback_history
-    assert len(opt.callback_history["null_cb"]) == (len(opt.opt_history) - 1) // 3
+    assert len(opt.callback_history["null_cb"]) == (len(opt.opt_history) + 2) // 3
```
Comment (zeyueN) [inline tests/test_training/test_opt.py:111]: I was expecting to have the `-1` removed, not replaced with a `+2`. There are a couple places callbacks depends on `len(opt_history)` , I'll take a look

Comment (zeyueN) [reply tests/test_training/test_opt.py:111]: I think we should now have an extra `+1` on [this line](https://github.com/XanaduAI/MrMustard/blob/a4ace7c6fd838dd3c2e76d79cf0ed764bc012d37/mrmustard/training/callbacks.py#L118).

Comment (ziofil) [reply tests/test_training/test_opt.py:111]: Thanks!

File: mrmustard/training/optimizer.py [hunk 2]
```diff
@@ -212,13 +215,14 @@ def compute_loss_and_gradients(cost_fn: Callable, parameters: List[Parameter]):
...
         if max_steps != 0 and len(self.opt_history) > max_steps:
             return True
-        if len(self.opt_history) > 20:  # if cost varies less than 10e-6 over 20 steps
-            if (
-                sum(abs(self.opt_history[-i - 1] - self.opt_history[-i]) for i in range(1, 20))
-                < 1e-6
-            ):
+        self.__history_sum += abs(self.opt_history[-2] - self.opt_history[-1])
+        if len(self.opt_history) > 20:  # stop if cost varies less than 10e-6 over 20 steps
+            self.__history_sum -= abs(self.opt_history[-21] - self.opt_history[-20])
```
Comment (ryk-wolf) [inline mrmustard/training/optimizer.py:224]: I can totally be mistaken, but it seems to me here that you're only looking at... 1 element? this:
`self.opt_history[-21] - self.opt_history[-20]`
is just 1 element afaik
is that what we want to be doing here?
I'm confused when seeing this line and this comment which mentions 20 steps
`# stop if cost varies less than 10e-6 over 20 steps`
Maybe I'm just misreading

File: mrmustard/training/callbacks.py [hunk 3]
```diff
@@ -115,7 +117,7 @@ def __post_init__(self):
 
     def get_opt_step(self, optimizer, **kwargs):  # pylint: disable=unused-argument
```
Comment (ryk-wolf) [inline mrmustard/training/callbacks.py:118]: As i see it, this function is doing 2 jobs: 
- setting self.optimizer_step
- returning optimizer_step

why not simply replace it by a computed property? and getting rid of the function.
Eg:
class Person():
    def __init__(self):
        self.n = "John Snow"
        self.age = 50
        self.is_old = False if self.age < 25 else True #computed property

**Review by ryk-wolf — COMMENTED** [review]
Looking good!

I made a few style comments, nothing blocking.

File: mrmustard/__init__.py [hunk 4]
```diff
@@ -33,9 +38,10 @@ def __new__(cls):  # singleton
     def __init__(self):
         self._backend = "tensorflow"
         self.HBAR = 2.0
-        self.CHOI_R = 0.881373587019543  # np.arcsinh(1.0)
         self.DEBUG = False
-        self.AUTOCUTOFF_PROBABILITY = 0.999  # capture at least 99.9% of the probability
+        self.AUTOCUTOFF_PROBABILITY = (
+            0.9999  # capture at least 99.99% of the probability by default
```
[the remark this answers is not in the document]
Comment (ziofil) [reply mrmustard/__init__.py:43]: my editor breaks long lines automatically

File: mrmustard/training/optimizer.py (hunk 2 above)
Comment (ziofil) [reply mrmustard/training/optimizer.py:224]: You are misreading. We maintain the difference of the last 20 steps: in line 222 we add the last difference and if opt_history is > 20 in line 224 we remove the difference beyond 20

File: mrmustard/utils/graphics.py [hunk 5]
```diff
@@ -14,13 +14,19 @@
...
 
+# pylint: disable=wrong-import-order
+
 from typing import Tuple
-from rich.progress import Progress, TextColumn, BarColumn, TimeRemainingColumn
+
 import matplotlib.pyplot as plt
-from matplotlib import cm
 import numpy as np
+from matplotlib import cm
```
[the remark this answers is not in the document]
Comment (ziofil) [reply mrmustard/utils/graphics.py:23]: I can't see the issue: both pyplot and cm are imported from matplotlib

File: tests/test_training/test_opt.py
[the remark this answers is not in the document]
Comment (ziofil) [reply tests/test_training/test_opt.py:1]: it's think it's pretty standard to have names begin with "test" (it also helps automated test discovery in vscode). But yes we could have more descriptive names.

File: mrmustard/training/callbacks.py (hunk 3 above)
Comment (ziofil) [reply mrmustard/training/callbacks.py:118]: yes it looks like it could be replaced by a property.

**Pull request closed without merging**